### PR TITLE
Other vector layers

### PR DESCRIPTION
**Pull request by llienher**
Tests I did and result:

URL only
 - roadmap medium -> OK
 - roadmap uuid -> OK
 - topo medium -> NOK (URL browser limit > 4094 characters)
 - topo uuid -> OK

LS only:
 - roadmap medium -> OK
 - roadmap uuid -> OK
 - topo medium -> OK
 - topo uuid -> OK

DB only:
 - roadmap medium -> OK
 - roadmap uuid -> OK
 - topo medium -> OK
 - topo uuid -> OK

Load URL, LS different
 - roadmap -> OK
 - topo -> OK

Load URL, DB different
 - roadmap -> OK
 - topo -> OK

Login with DB -> OK
Logout with DB -> OK

### CODE REVIEW

**Comment by tonio** [comment]
Small comment before starting review/test: I think this PR should target `c2cnextprod` branch (-:

**Comment by llienher** [comment]
Branch rebased on `c2cnextprod`.

File: geoportal/geoportailv3_geoportal/static-ngeo/js/apps/MainController.js [hunk 1]
```diff
@@ -1071,16 +1210,26 @@ const MainController = function(
...
+      this.appMvtStylingService.saveStyle(dataObject, isPublished).then(id => {
+        const isValidUUIDv4Regex = /^[0-9A-F]{8}-[0-9A-F]{4}-4[0-9A-F]{3}-[89AB][0-9A-F]{3}-[0-9A-F]{12}$/gi;
+          // If result is a serialized UUID
+          if (id.match(isValidUUIDv4Regex) !== null) {
+            this.ngeoLocation_.updateParams({
+              'serial': id,
+              'serialLayer': bgLayer.get('label')
+            });
+            this.ngeoLocation_.refresh();
+          }
```
[the remark this answers is not in the document]
Comment (llienher) [reply geoportal/geoportailv3_geoportal/static-ngeo/js/apps/MainController.js]: Yep, fixed.

File: geoportal/geoportailv3_geoportal/static-ngeo/js/apps/MainController.js [hunk 2]
```diff
@@ -234,6 +241,62 @@ function getDefaultMediumStyling() {
...
+    symbols: ['lu_contour-label-100','lu_contour-label-20','lu_apex_label'],
+    visible: true
+  }, {
+    label: gettext('Hillshade'),
+    hillshades: ['lu_hillshade'],
+    visible: true
+  }
+];
+}
+
```
Comment (tonio) [inline geoportal/geoportailv3_geoportal/static-ngeo/js/apps/MainController.js:299]: Maybe we consider having these verbose config in another than the already too big MainController.

File: geoportal/geoportailv3_geoportal/static-ngeo/js/apps/MainController.js [hunk 3]
```diff
@@ -324,46 +387,63 @@ const MainController = function(
...
-      mbMap.setPaintProperty(path, 'line-color', item.color);
-      mbMap.setPaintProperty(path, 'line-opacity', 1);
+      if (item.color) {
+        mbMap.setPaintProperty(path, 'line-color', item.color);
+        mbMap.setPaintProperty(path, 'line-opacity', 1);
+      }
+      mbMap.setLayoutProperty(path, 'visibility', item.visible ? 'visible' : 'none');
+    });
+    (item.symbols || []).forEach(path => {
+      if(item.color) {
```
Comment (tonio) [inline geoportal/geoportailv3_geoportal/static-ngeo/js/apps/MainController.js:407]: useless test? (I mean not used inside following block)

File: geoportal/geoportailv3_geoportal/static-ngeo/js/apps/MainController.js [hunk 4]
```diff
@@ -386,60 +466,62 @@ const MainController = function(
...
 
-  const mediumStyle = appMvtStylingService.getMediumStyle();
+  /* const mediumStyle = appMvtStylingService.getStyle();
   if (mediumStyle !== undefined) {
     mediumStyle.then((style) => {
         Object.assign(this.mediumStylingData, JSON.parse(style || '{}'));
         this.checkSelectedSimpleData();
       });
-  }
+  } */
```
Comment (tonio) [inline geoportal/geoportailv3_geoportal/static-ngeo/js/apps/MainController.js:518]: Better remove than comment?

File: geoportal/geoportailv3_geoportal/static-ngeo/js/apps/MainController.js [hunk 5]
```diff
@@ -903,6 +990,55 @@ const MainController = function(
...
+        this.isColorVisible = true;
+        $('#editor-medium').collapse('hide');
+      } else {
+        this.isColorVisible = false;
+        $('#editor-medium').collapse('show');
+      }
+
+      // Only if current is a vector tiles layer
+      // Check if it is a function, otherwise it sould be "is not a function" error
+      if (typeof current.getMapBoxMap === 'function') {
```
Comment (tonio) [inline geoportal/geoportailv3_geoportal/static-ngeo/js/apps/MainController.js:1012]: Can you have something else than `function` or `undefined` ? can't `if (current.getMapBoxMap){` be enough?

File: geoportal/geoportailv3_geoportal/static-ngeo/js/apps/MainController.js [hunk 6]
```diff
@@ -1071,16 +1211,26 @@ const MainController = function(
...
-        this.appMvtStylingService.removeMediumStyle();
+      const dataObject = {
+        medium: undefined, // empty for the save function
+        background: bgLayer
+      }
+
+      this.appMvtStylingService.saveStyle(dataObject, isPublished).then(id => {
+        const isValidUUIDv4Regex = /^[0-9A-F]{8}-[0-9A-F]{4}-4[0-9A-F]{3}-[89AB][0-9A-F]{3}-[0-9A-F]{12}$/gi;
+        // If result is a serialized UUID
+        if (id.match(isValidUUIDv4Regex) !== null) {
```
Comment (tonio) [inline geoportal/geoportailv3_geoportal/static-ngeo/js/apps/MainController.js:1223]: What about:
```suggestion
        if (isValidUUIDv4Regex.test(id)) {
```

File: geoportal/geoportailv3_geoportal/static-ngeo/js/mvtstyling/MvtStylingService.js [hunk 7]
```diff
@@ -8,357 +9,391 @@ const url_get = '/get_userconfig';
...
+            keyword = 'roadmap';
+            break;
+        case 'topogr_global':
+            keyword = 'topomap';
+            break;
+        case 'topo_bw_jpeg':
+            keyword = 'topomap_gray';
+            break;
+    }
+    return keyword;
```
Comment (tonio) [inline geoportal/geoportailv3_geoportal/static-ngeo/js/mvtstyling/MvtStylingService.js:38]:
```suggestion
      return {
        'basemap_2015_global': 'roadmap',
        'topogr_global': 'topomap';
         'topo_bw_jpeg': 'topomap_gray';
      }[label]
```

File: geoportal/geoportailv3_geoportal/static-ngeo/js/mvtstyling/MvtStylingService.js [hunk 8]
```diff
@@ -8,357 +9,391 @@ const url_get = '/get_userconfig';
...
+        this.ngeoLocation_ = ngeoLocation;
+    }
+
+    getBgStyle() {
+        const layerLabelList = ['basemap_2015_global', 'topogr_global', 'topo_bw_jpeg'];
+        const promises = [];
+        layerLabelList.forEach(label => {
+            const promise = this.setConfigForLayer_(label);
+            promises.push(promise);
+        });
```
Comment (tonio) [inline geoportal/geoportailv3_geoportal/static-ngeo/js/mvtstyling/MvtStylingService.js:74]:
```suggestion
        layerLabelList.forEach(label => promises.push( this.setConfigForLayer_(label)));
```

File: geoportal/geoportailv3_geoportal/static-ngeo/js/mvtstyling/MvtStylingService.js [hunk 9]
```diff
@@ -8,357 +9,391 @@ const url_get = '/get_userconfig';
...
+            xyz: xyz_custom || defaultMapBoxStyleXYZ,
+            xyz_custom,
+            style: defaultMapBoxStyle
+        };
+
+        const serial = new URLSearchParams(window.location.search).get('serial');
+        const serialLayer = new URLSearchParams(window.location.search).get('serialLayer');
+        if (serial) {
+            // if serial is number id, retrieve style form it
+            const isValidUUIDv4Regex = /^[0-9A-F]{8}-[0-9A-F]{4}-4[0-9A-F]{3}-[89AB][0-9A-F]{3}-[0-9A-F]{12}$/gi;
```
Comment (tonio) [inline geoportal/geoportailv3_geoportal/static-ngeo/js/mvtstyling/MvtStylingService.js:102]: this regexp shouldn't be duplicated.

File: geoportal/geoportailv3_geoportal/static-ngeo/js/mvtstyling/MvtStylingService.js [hunk 10]
```diff
@@ -8,357 +9,391 @@ const url_get = '/get_userconfig';
...
+            if (serial.match(isValidUUIDv4Regex) !== null) {
```
Comment (tonio) [inline geoportal/geoportailv3_geoportal/static-ngeo/js/mvtstyling/MvtStylingService.js:103]: same as above

File: geoportal/geoportailv3_geoportal/static-ngeo/js/mvtstyling/MvtStylingService.js [hunk 11]
```diff
@@ -8,357 +9,391 @@ const url_get = '/get_userconfig';
...
-                      this.backgroundLayerMgr_.set(map, bgLayer);
-                    });
-                } catch (error) {
-                    console.log(error);
-                    return;
-                }
-            }, 50);
+    publishIfSerial(map) {
+        const serial = new URLSearchParams(window.location.search).get('serial');
+        const isValidUUIDv4Regex = /^[0-9A-F]{8}-[0-9A-F]{4}-4[0-9A-F]{3}-[89AB][0-9A-F]{3}-[0-9A-F]{12}$/gi;
```
Comment (tonio) [inline geoportal/geoportailv3_geoportal/static-ngeo/js/mvtstyling/MvtStylingService.js:186]: again here

File: geoportal/geoportailv3_geoportal/static-ngeo/js/mvtstyling/MvtStylingService.js [hunk 12]
```diff
@@ -8,357 +9,391 @@ const url_get = '/get_userconfig';
...
+        if (serial) {
+            // check if simple/medium styling
+            if (serial.match(isValidUUIDv4Regex) === null) {
+                const bgLayer = this.backgroundLayerMgr_.get(map);
+                const interval = setInterval(() => {
```
Comment (tonio) [inline geoportal/geoportailv3_geoportal/static-ngeo/js/mvtstyling/MvtStylingService.js:191]: shouldn't it be just a setTimeout ? doesn't look repeated.

File: geoportal/geoportailv3_geoportal/static-ngeo/js/mvtstyling/MvtStylingService.js [hunk 13]
```diff
@@ -8,357 +9,391 @@ const url_get = '/get_userconfig';
...
-    const isValidUUIDv4Regex = /^[0-9A-F]{8}-[0-9A-F]{4}-4[0-9A-F]{3}-[89AB][0-9A-F]{3}-[0-9A-F]{12}$/gi;
-    if (serial) {
-        // check if simple/medium styling
-        if (serial.match(isValidUUIDv4Regex) === null) {
-            const bgLayer = this.backgroundLayerMgr_.get(map);
-            if (!bgLayer) { // 3D enable watch is run before bgLayer is initialized...
-                return;
+    unpublishIfSerial(map) {
+        const serial = new URLSearchParams(window.location.search).get('serial');
+        const isValidUUIDv4Regex = /^[0-9A-F]{8}-[0-9A-F]{4}-4[0-9A-F]{3}-[89AB][0-9A-F]{3}-[0-9A-F]{12}$/gi;
```
Comment (tonio) [inline geoportal/geoportailv3_geoportal/static-ngeo/js/mvtstyling/MvtStylingService.js:215]: (-:

File: geoportal/geoportailv3_geoportal/static-ngeo/js/mvtstyling/MvtStylingService.js [hunk 14]
```diff
@@ -8,357 +9,391 @@ const url_get = '/get_userconfig';
...
+        if (serial) {
+            // check if simple/medium styling
+            if (serial.match(isValidUUIDv4Regex) === null) {
+                const bgLayer = this.backgroundLayerMgr_.get(map);
+                if (!bgLayer) { // 3D enable watch is run before bgLayer is initialized...
+                    return;
+                }
+                return this.unpublishStyle(bgLayer);
```
Comment (tonio) [inline geoportal/geoportailv3_geoportal/static-ngeo/js/mvtstyling/MvtStylingService.js:223]: I think should 3 nested `if` could be simplified for readability
(eg: `if (!serial && !isValid.test(serial)) return;`.

**Review by tonio — COMMENTED** [review]
Looks to works pretty well, but url limit is reached very often to be fully testable.
Do we need to discuss this point or merge & let them test?

Otherwise, a few minor comments, mainly code simplification for readability.

File: geoportal/geoportailv3_geoportal/static-ngeo/js/apps/MainController.js (hunk 2 above)
Comment (llienher) [reply geoportal/geoportailv3_geoportal/static-ngeo/js/apps/MainController.js:300]: I guess we could move it, but for now I don't know where so it is a minor change for the future.

**Pull request merged**